libs/core/src/repo/schema_v3.rs

```diff
@@ -0,0 +1,123 @@
+use crate::CoreError;
+use hmdb::transaction::Transaction;
+use lockbook_shared::account::Account;
+use lockbook_shared::file_like::FileLike;
+use lockbook_shared::file_metadata::Owner;
+use lockbook_shared::signed_file::SignedFile;
+use lockbook_shared::{document_repo, SharedError};
+use serde::{Deserialize, Serialize};
+use std::fs;
+use uuid::Uuid;
+
+use super::schema_v2::CoreV2;
+
+pub type Tx<'a> = transaction::CoreV3<'a>;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OneKey;
+
+hmdb::schema! {
+    CoreV3 {
+        account: <OneKey, Account>,
+        last_synced: <OneKey, i64>,
+        root: <OneKey, Uuid>,
+        local_metadata: <Uuid, SignedFile>,
+        base_metadata: <Uuid, SignedFile>,
+        read_activity: <Uuid, usize>,
+        write_activity: <Uuid, usize>,
+        public_key_by_username: <String, Owner>,
+        username_by_public_key: <Owner, String>
+    }
+}
+
+impl CoreV3 {
+    pub fn init_with_migration(writeable_path: &str) -> Result<CoreV3, CoreError> {
+        let db = CoreV3::init(writeable_path)?;
+
+        // migrate metadata from v1
+        if db.account.get_all()?.is_empty() {
+            let source = CoreV2::init(writeable_path)?;
+            db.transaction(|tx| {
+                // copy existing tables (new tables populated on next sync)
+                for v in source.account.get_all()?.into_values() {
+                    tx.account.insert(OneKey {}, v);
+                }
+                for v in source.last_synced.get_all()?.into_values() {
+                    tx.last_synced.insert(OneKey {}, v);
+                }
+                for v in source.root.get_all()?.into_values() {
+                    tx.root.insert(OneKey {}, v);
+                }
+                for (k, v) in source.local_metadata.get_all()? {
+                    tx.local_metadata.insert(k, v);
+                }
+                for (k, v) in source.base_metadata.get_all()? {
+                    tx.base_metadata.insert(k, v);
+                }
+                for (k, v) in source.public_key_by_username.get_all()? {
+                    tx.public_key_by_username.insert(k, v);
+                }
+                for (k, v) in source.username_by_public_key.get_all()? {
+                    tx.username_by_public_key.insert(k, v);
+                }
+                Ok::<_, hmdb::errors::Error>(())
+            })
+            .expect("failed to migrate local database from v2 to v3")?;
+        }
+
+        // migrate documents from id+source structure to id+hmac structure
```

Comment:
still a wip I assume because you didn't actually solicit a review, but v2 -> v3 will likely not need any of this stuff